rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,364 @@
+It will be versioned separately from Teleport, and manage the installation of the correct Teleport agent version manually. 
+It will read the unauthenticated `/v1/webapi/ping` endpoint from the Teleport proxy, parse new fields on that endpoint, and install the specified agent version according to the specified upgrade plan.
+It will download the correct version of Teleport as a tarball, unpack it in `/var/lib/teleport`, and ensure it is symlinked from `/usr/local/bin`.
+
+### Installation
+
+```shell
+$ apt-get install teleport-ent-updater
+$ teleport-update enable --proxy example.teleport.sh
+
+# if not enabled already, configure teleport and:
+$ systemctl enable teleport
+```
+
+### API
+
+#### Endpoints
+
+`/v1/webapi/ping`
+```json
+{
+  "server_edition": "enterprise",
+  "agent_version": "15.1.1",
+  "agent_auto_update": true,
+  "agent_update_after": "2024-04-23T18:00:00.000Z",
+  "agent_update_jitter_seconds": 10,
+}
+```
+Notes:
+- Critical updates are achieved by serving `agent_update_after` with the current time.
+- The Teleport proxy translates upgrade hours (below) into a specific time after which all agents should be upgraded.
+- If an agent misses an upgrade window, it will always update immediately.
+- The edition served is the cluster edition (enterprise, enterprise-fips, or oss), and cannot be configured.
+
+#### Teleport Resources
+
+```yaml
+kind: cluster_maintenance_config
+spec:
+  # agent_auto_update allows turning agent updates on or off at the
+  # cluster level. Only turn agent automatic updates off if self-managed
+  # agent updates are in place.
+  agent_auto_update: on|off
+  # agent_update_hour sets the hour in UTC at which clients should update their agents.
+  agent_update_hour: 0-23
+  # agent_update_now overrides agent_update_hour and sets agent update time to the current time.
+  # This is useful for rolling out critical security updates and bug fixes.
+  agent_update_now: on|off
+  # agent_update_jitter_seconds sets a duration in which the upgrade will occur after the hour.
+  # The agent upgrader will pick a random time within this duration in which to upgrade.
+  agent_update_jitter_seconds: 0-MAXINT64
+  
+  [...]
+```
+```
+$ tctl autoupdate update --set-agent-auto-update=off
+Automatic updates configuration has been updated.
+$ tctl autoupdate update --set-agent-update-hour=3
+Automatic updates configuration has been updated.
+$ tctl autoupdate update --set-agent-update-now=true
+Automatic updates configuration has been updated.
+$ tctl autoupdate update --set-agent-update-jitter-seconds=600
+Automatic updates configuration has been updated.
+```
+
+```yaml
+kind: autoupdate_version
+spec:
+  # agent_version is the version of the agent the cluster will advertise.
+  # Can be auto (match the version of the proxy) or an exact semver formatted
+  # version.
+  agent_version: auto|X.Y.Z
+
+  [...]
+```
+```
+$ tctl autoupdate update --set-agent-version=15.1.1
```

Comment:
Would it be possible to allow cloud customers to configure this? It would allow us to safely opt into versions at our own pace (until we can bucket instances into PROD/PREPROD etc)

Reply:
Cloud performs cluster upgrades, and we'd like to ensure that the cluster and agent versions are always extensively tested and compatible. Uncoordinated cluster and agent upgrades could lead to incompatible versions and lost access to resources.

> bucket instances into PROD/PREPROD etc

It sounds like you want to rollout new agent versions across the same cluster in buckets. Is this correct?

If so, I do have an idea for bucketed agent upgrades. It would look something like this:
- Add bucket IDs that are settable via `teleport-updater enable --bucket 2`
- Buckets time offsets would be user-configurable via `tctl` in `cluster_maintenance_config`
- `/v1/webapi/ping` would return separate times for each bucket

Would this work for your use case?

I may open a separate RFD to add this in the future (depending on user feedback)

Reply:
> It sounds like you want to rollout new agent versions across the same cluster in buckets. Is this correct?

That is correct yes.

> Would this work for your use case?

It can work (`stable` and `latest` buckets would be sufficient for us TBH), where all of our PREPROD agents are enrolled into `latest` and our PROD into `stable`. This is currently the system we use with our custom Teleport version server

However due to past issues with Teleport agent auto-upgrading, we would not be able to onboard onto this new approach unless we have the ability to control the roll-out of updates to our agents. Either as defined above, or something akin to https://github.com/gravitational/teleport/pull/40190#discussion_r1568375714, while the approach defined above allows for buckets